addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MavenMetadataGenerator.java

```diff
@@ -517,15 +526,111 @@ private Metadata generateGroupMetadata( final Group group, final List<ArtifactSt
 
         if ( master != null )
         {
-            Map<String, MetadataInfo> cacheMap = new HashMap<>();
-            cacheMap.put( toMergePath, new MetadataInfo( master ) );
-            versionMetadataCache.put( group.getKey(), cacheMap );
+            putToMetadataCache( group.getKey(), toMergePath, master );
             return master;
         }
 
         return null;
     }
 
+    private void putToMetadataCache( StoreKey key, String toMergePath, Metadata meta )
+    {
+        synchronized ( versionMetadataCache )
+        {
+            Map cacheMap = versionMetadataCache.get( key );
```

Comment:
Does this have type params? We should implement a putIfAbsent() call for CacheHandle...

Reply:
No type param.